### PR TITLE
[Hardware][TPU]Enable ragged paged attention kernel and resolve recompilation issue

### PR DESCRIPTION
**Pull request by yaochengji**
- enable ragged paged attention kernel
- resolve recompilation issue
- change the layout of kv cache from (num_kv_heads, num_blocks, block_size, head_size) to (num_blocks, block_size, num_kv_heads, head_size)
- update torch xla nightly wheel

### CODE REVIEW

**Comment by yaochengji** [comment]
This PR depends on https://github.com/pytorch/xla/pull/8797, I will update the torch/xla nightly version tomorrow

File: vllm/v1/attention/backends/pallas.py [hunk 1]
```diff
@@ -168,8 +164,9 @@ def forward(
             attn_metadata.num_seqs,
             num_kv_pages_per_block=NUM_KV_PAGES_PER_BLOCK,
             num_queries_per_block=NUM_QUERIES_PER_BLOCK,
-            use_kernel=use_kernel,
-        )
+            vmem_limit_bytes=32 * 1024 * 1024,
```
[the remark this answers is not in the document]
Comment (yaochengji) [reply vllm/v1/attention/backends/pallas.py:167]: Yes, @vanbasten23 is tuning the kernel parameter, will update it later.

**Review by alexm-redhat — COMMENTED** [review]
@yaochengji thanks for the PR! I did a first review pass and have some questions about parameters that would be good to understand.

File: vllm/v1/worker/tpu_model_runner.py [hunk 2]
```diff
@@ -428,23 +428,22 @@ def _prepare_inputs(self, scheduler_output: "SchedulerOutput"):
...
-                                                  padded_total_num_scheduled_tokens]
+        padded_block_table = self.block_table_cpu[:self.max_num_reqs]
         padded_block_table[:num_reqs, :self.max_num_blocks_per_req] = (
             self.input_batch.block_table.get_cpu_tensor()[:num_reqs])
         padded_block_table = padded_block_table.to(self.device)
-        query_start_loc = self.query_start_loc_cpu[:
-                                                   padded_total_num_scheduled_tokens
-                                                   + 1].to(self.device)
-        seq_lens = self.seq_lens_cpu[:padded_total_num_scheduled_tokens].to(
+        query_start_loc = self.query_start_loc_cpu[:self.max_num_reqs + 1].to(
```
[the remark this answers is not in the document]
Comment (yaochengji) [reply vllm/v1/worker/tpu_model_runner.py:435]: Because the first dimension of block_table, query_start_loc  and seq_lens are at most `max_num_reqs`. Padding them based on token number is quite a waste.

File: vllm/v1/worker/tpu_model_runner.py [hunk 3]
```diff
@@ -693,29 +692,34 @@ def _dummy_run(
...
                                    device=self.device)
         slot_mapping = torch.zeros(num_tokens,
                                    dtype=torch.int64,
                                    device=self.device)
-        block_tables = torch.zeros((num_tokens, self.block_table_cpu.shape[1]),
-                                   dtype=torch.int32,
-                                   device=self.device)
-        query_lens = [1] * num_tokens
+        block_tables = torch.zeros(
+            (self.max_num_reqs, self.block_table_cpu.shape[1]),
```
[the remark this answers is not in the document]
Comment (yaochengji) [reply vllm/v1/worker/tpu_model_runner.py:703]: To avoid recompilation, then only dimension that can vary is the `token_num`.

File: vllm/v1/worker/tpu_model_runner.py [hunk 4]
```diff
@@ -693,29 +692,34 @@ def _dummy_run(
...
+        num_seqs = torch.tensor([actual_num_reqs],
+                                dtype=torch.int32,
+                                device=self.device)
         attn_metadata = PallasMetadata(
             slot_mapping=slot_mapping,
             block_tables=block_tables,
             context_lens=context_lens,
             query_start_loc=query_start_loc,
-            num_seqs=num_tokens,
+            num_seqs=num_seqs,
```
[the remark this answers is not in the document]
Comment (yaochengji) [reply vllm/v1/worker/tpu_model_runner.py:722]: `actual_num_reqs` is int and `num_seqs` is Tensor. torch dynamo will always recompile if the value of an int variable changes.

**Comment by yaochengji** [comment]
@mgoin @alexm-redhat Thanks for your comments! The PyTorch/XLA nightly wheel is updated and now this PR is ready for review!

File: vllm/v1/attention/backends/pallas.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (yaochengji) [reply vllm/v1/attention/backends/pallas.py]: The parameter is updated, thanks!

File: vllm/v1/worker/tpu_model_runner.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (alexm-redhat) [reply vllm/v1/worker/tpu_model_runner.py:435]: oh I see, this is already maxed, ok makes sense.

File: vllm/v1/worker/tpu_model_runner.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (alexm-redhat) [reply vllm/v1/worker/tpu_model_runner.py:703]: ok cool

File: vllm/v1/worker/tpu_model_runner.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (alexm-redhat) [reply vllm/v1/worker/tpu_model_runner.py:722]: ok now I see it

**Review by alexm-redhat — APPROVED** [review]
@yaochengji LGTM! ready to merge when green.

**Review by mgoin — APPROVED** [review]
i was able to run gsm8k on qwen2, great work!
```
vllm (pretrained=Qwen/Qwen2-1.5B-Instruct,max_model_len=2048,trust_remote_code=True), gen_kwargs: (None), limit: None, num_fewshot: 5, batch_size: auto
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.5921|±  |0.0135|
|     |       |strict-match    |     5|exact_match|↑  |0.5861|±  |0.0136|
```
